### ISSUE TITLE
pihole/pihole:v5.8 won't start

### ISSUE THREAD

**Issue by kylekurz** [pi-hole/docker-pi-hole#834]
This is a: Bug

## Details

In an attempt to vet the PR you merged for me earlier today, I pulled in the latest pihole/pihole image and attempted to run on two different machines, both failed with the following error:

`  [✗] DNS service is NOT listening
kill: usage: kill [-s sigspec | -n signum | -sigspec] pid | jobspec ... or kill -l [sigspec]`

## Related Issues
- [ x] I have searched this repository/Pi-hole forums for existing issues and pull requests that look similar 

## How to reproduce the issue 

1. Environment data
  * Operating System: Raspbian
  * Hardware: Pi 4B and Pi 3B+ 
  * Kernel Architecture: 
  * Docker Install Info and version: 
    - Software source: Official docker-ce `Docker version 20.10.5, build 55c4c88` 
    - Supplimentary Software: `docker-compose version 1.29.0, build unknown` 
  * Hardware architecture: `Linux pihole 5.10.17-v7l+ #1403 SMP Mon Feb 22 11:33:35 GMT 2021 armv7l GNU/Linux` 

2. docker-compose.yml contents, docker run shell command, or paste a screenshot of any UI based configuration of containers here:
Here's my compose file that works if I go back to version 5.2.4 that I had from before the upgrade:
```
version: "3"
services:
        pihole:
                container_name: pihole
                domainname: docker
                hostname: pihole
                image: pihole/pihole:v5.8
                network_mode: host
                dns:
                        - 127.0.0.1
                        - 1.1.1.1
                restart: unless-stopped
                volumes:
                        - /home/pi/docker-compose/pihole/docker/pihole/pihole:/etc/pihole
                cap_add:
                        - NET_ADMIN
                environment:
                        - ServerIP=192.168.25.5
                        - TZ=PST
                        - WEBPASSWORD=changeme
                        - DNS1='127.0.0.1#5053'
                        - DNS2=''

        cloudflared:
                container_name: cloudflared
                domainname: docker
                image: visibilityspots/cloudflared:alpine-3.12
                network_mode: host
                restart: unless-stopped
                environment:
                        - PORT=5053
```
3. any additional info to help reproduce
I have a more complex setup involving ansible and docker swarm/stack that I'm attempting to deploy, but I backed down to this to try to remove my issues as much as possible and limit it to the new container.

## These common fixes didn't work for my issue

- [ x] I have tried removing/destroying my container, and re-creating a new container
- [ x] I have tried fresh volume data by backing up and moving/removing the old volume data
- [ ] I have tried running the stock `docker run` example(s) in the readme (removing any customizations I added)
- [x ] I have tried a newer or older version of Docker Pi-hole (depending what version the issue started in for me)
- [ ] I have tried running without my volume data mounts to eliminate volumes as the cause

If the above debugging / fixes revealed any new information note it here.
Add any other debugging steps you've taken or theories on root cause that may help.

**Comment by rmartin16** [comment]
FWIW....i run three HA docker pi-holes....the primary one didn't came back up after the upgrade with the problems you detail.

I finally rebuilt the volumes and finally got it working again.....i couldn't find the root cause for the life of me, though...

**Comment by kylekurz** [comment]
I removed the entire local persistent store and still couldn't get it running through stack/swarm. I'll duplicate my docker-compose backup, clear out the local storage and try that.

**Comment by kylekurz** [comment]
Same behavior for me on cleaned directory with the listed compose file.

**Comment by GnaphronG** [comment]
I've got the same problem with pihole 5.8 in docker on Raspberry pi3, I managed to get it started using 
`docker-compose exec pihole pihole-FTL` once the start process got stuck in the dns not starting

**Comment by humsg** [comment]
after many retries, I got this: edit nameserver (etc/resolv.conf or /mnt/dietpi_userdata/docker-data/containers/(random id)/resolv.conf) to public dns, then restart pihole can fix the problem, but I do not think this is the good way to go. I even reset everything then I realized that it is not my problem. sad
I am on docker with unbound up and running with not problem with previous dockered pihole ver
and I do still got 
  [✗] DNS service is NOT listening 
after pihole restart, but everything with unbound seems running properly.

I would like the nameserver runs as before, in docker 172.17.0.1 not public dns. thx

**Comment by humsg** [comment]
so the problem seems like it's getting looped with my original settings which was router to Pi4B, Pi4B to docker(172.17.0.1), docker container(pihole and unbound) to docker(172.17.0.1), dockered pihole to dockered unbound. I was not notice how it was working, because I had no problem with v5.2.4 v5.4 v5.7 or even before. I want everything looped, unbound is the only way out.

**Comment by GnaphronG** [comment]
Thank you @humsg , you gave me the idea the invert the dns in docker-compose.yml
```
  dns:                                                                                       
      - 1.1.1.1                                
      - 127.0.0.1   
 ```
Now is get 
```
pihole         |   [✗] DNS service is NOT listening                                            
pihole         | kill: usage: kill [-s sigspec | -n signum | -sigspec] pid | jobspec ... or kill -l [sigspec]
```
but it seems to not bother it so much and the services are up and running

**Comment by uselpa** [comment]
Same issues on x86. Had to revert back to v5.7.

**Comment by PromoFaux** [comment]
All, just a hunch.. what is the output of docker --version on your affected machines?

**Comment by predmijat** [comment]
Docker version 20.10.6, build 370c28948e

**Comment by clay9247** [comment]
Docker version 20.10.5, build 55c4c88

**Comment by PromoFaux** [comment]
Yeah, interestingly I cannot reproduce this on my synology NAS (running docker 18.09.8) but on windows, running docker 20.10.5, I cannot even get a 5.7 container to start. which makes me wonder if something _other_ has happened in docker 20....

If I recall correctly, I saw an update for docker on windows a few days ago

**Comment by clay9247** [comment]
Interestingly reverting to 5.7 did not work for me either. I got stuck on [✗] DNS resolution is currently unavailable and ([✗] DNS service is not running on 5.8). FTL was not starting. With providing 1.1.1.1 as primary dns server to the container it started to pull the blocklists and started FTL but promted "[✗] DNS service is NOT listening
pihole    | kill: usage: kill [-s sigspec | -n signum | -sigspec] pid | jobspec ... or kill -l [sigspec]". DNS resolution however seems to work fine now with unbound being the only DNS server.

**Comment by PromoFaux** [comment]
the `kill: usage...` message is just cosmetic at this stage (and a reminder for us to sort that part out)

**Comment by predmijat** [comment]
Reverting to 5.7 works for me. I have this in my docker-compose.yml:
```
    environment:
      DNS1: 1.1.1.1
      DNS2: 8.8.8.8
```

**Comment by PromoFaux** [comment]
the DNS1 and DNS2 environment variables are deprecated, you should use `PIHOLE_DNS_ : 1.1.1.1;8.8.8.8`, however those are for the upstream servers that FTL forwards queries from devices on to and not what it uses itself, if that makes any sense?

**Comment by kvn1351** [comment]
Watchtower just updated my pihole container. And thus all my devices stopped being able to resolve DNS requests...
Returning to 5.7 fixed the problem immediately. In my opinion this is not ok guys.. A 5.X release shouldn't break a setup completely.

For the record, it hangs on:

```
 ::: Docker start setup complete,
  [✗] DNS service is not running,
```

**Comment by PromoFaux** [comment]
Personal opinion : I would _never_ use watchtower to automatically update something as critical to my network as a DNS server. In an ideal world things wouldn't break, but sometimes they do (we're only human after all)  - and so you need to be around to immediately roll back in case of errors. I can't imagine I would want anything like that to update over night when I'm not there. As a developer myself - I know not to fully trust other developers 😉

At the moment I'm struggling to reproduce this consistently. Hate to use the "Works on my machine" excuse... but before tagging v5.8 - I (manually 😉) updated my system and everything worked... so it was tagged.

It's a bit beyond me, to be honest, what is going wrong here. The guy who usually maintains this image has practically gone AWOL, so I've been trying my best to get updates out as we update the core Pi-hole packages, but I'm no docker expert

**Comment by kvn1351** [comment]
> Personal opinion : I would _never_ use watchtower to automatically update something as critical to my network as a DNS server.

You're 100% right. In fact, I had just setup my pihole instances a few days ago and forgot to set the tag to a fixed version. In a way, I'm lucky that this has happened so soon.

Thanks for all of your work btw, I'm not here to shit on it. Mistakes happen, especially when it's seemingly working fine on your end.

**Comment by kylekurz** [comment]
> Watchtower just updated my pihole container. And thus all my devices stopped being able to resolve DNS requests...
> Returning to 5.7 fixed the problem immediately. In my opinion this is not ok guys.. A 5.X release shouldn't break a setup completely.
>
> Additional info:
> 
> * Linux 5.11.7-1-pve x86_64
> * Docker version 18.09.1, build 4c52b90

While I understand your frustration, let’s please keep comments relevant to finding and resolving the issue. No one is saying “breaking on a point release is acceptable”. 

@PromoFaux one thing in this comment that I did notice, this user is on docker 18. So maybe digging into docker versions as a cause isn’t useful. I’ll try to dig in a little bit more later this evening if I have time.

**Comment by PromoFaux** [comment]
> So maybe digging into docker versions as a cause isn’t useful

More than possibly correct, at this stage I'm clutching at straws 😏

**Comment by squirtbrnr** [comment]
For those saying v5.8 works, what is the system DNS for Pi-hole? I only had `--dns=127.0.0.1` And that’s when I got the DNS service NOT started. Once I added `--dns=1.1.1.1` It would start but then I get DNS service NOT listening and I had to click Enable from the PiHole webpage. I also have my network setting on my Synology NAS manually configured for a static IP and the DNS points to my router which then the router points the NAS right back at itself for DNS. I’m grasping at straws because I’m not a network tech, but is it possible there is some kind of DNS loop occurring between PiHole and the NAS/Router? And for the record, v5.7 works for me, but v5.8 does not.

**Comment by lightswitch05** [comment]
Its working for me too, on a NUC with Ubuntu 20.04.2 LTS. Docker version `Docker version 20.10.5, build 55c4c88`. I'm happy to share my [docker-compose.yaml](https://gist.github.com/lightswitch05/cb0087d2f0a18a110d0e276a43c39b57)

**Comment by PromoFaux** [comment]
FWIW - here is my working `docker-compose.yml` file 

```
bash-4.4# docker --version
Docker version 18.09.8, build bfed4f5
bash-4.4# docker-compose --version
docker-compose version 1.24.0, build 0aa59064
```

```yml
version: '3.5'
services:

  pihole:
    container_name: pihole
    hostname: pihole-docker
    depends_on:
      - unbound
    image: pihole/pihole:v5.8
    environment:
      - TZ=Europe/London
      - ServerIP=192.168.1.253
      - REV_SERVER=true
      - REV_SERVER_DOMAIN=lan
      - REV_SERVER_TARGET=192.168.0.1
      - REV_SERVER_CIDR=192.168.0.0/16
      - SKIPGRAVITYONBOOT=1
      - WEBPASSWORD=test
    volumes:
      - ./pihole/configs/pihole/:/etc/pihole/
      - ./pihole/configs/dnsmasq.d:/etc/dnsmasq.d/
    mac_address: d0:ca:ab:cd:ef:fe
    dns:
     - 8.8.8.8
    networks:
      home:
        ipv4_address: 192.168.1.253
    restart: always
```

**Comment by diginc** [comment]
The only way I was able to reproduce any issue was running a single localhost DNS server in the below style, which I think a group of people has an issue with an update or two ago also.

```
    environment:
      PH_VERBOSE: 1
    dns:
      - 127.0.0.1

... equals ...

pihole         | + gravity.sh
pihole         |   [✗] DNS service is not running
Looping until failure
```

I think solo localhost dns only works if you have the environment variable `SKIPGRAVITYONBOOT: 1` .  

So that^ or setting a secondary DNS should work for most.

**Comment by PromoFaux** [comment]
All, talking with @diginc offline - try removing the `--dns`  flag from your compose rules / run commands and seeing if that makes a difference?

Per the [quickstart](https://github.com/pi-hole/docker-pi-hole#quick-start) in the Readme, no `dns` setting is set there - but we do mention it below for `v4.1.1+` (we may need to revise that as it looks like it's causing issues)

**Comment by kylekurz** [comment]
@PromoFaux on my stack/swarm deploy, I have no `dns` set at all and still had the same error. I will try with compose.

**Comment by kylekurz** [comment]
This compose still fails:

```
version: "3.1"
services:
  pihole:
    hostname: pihole
    image: pihole/pihole:v5.8
    depends_on:
      - cloudflared
    volumes:
      - /home/pi/docker-compose/pihole/docker/pihole/pihole:/etc/pihole
    environment:
      - ServerIP=192.168.25.233
      - TZ=PST
      - PIHOLE_DNS_='192.168.25.233#5054'
    ports:
      - "53:53/tcp"
      - "53:53/udp"
      - "67:67/udp"
      - "80:80/tcp"

  cloudflared:
    image: visibilityspots/cloudflared:alpine-3.12
    ports:
      - "5054:5054/udp"
```

Here's the entirety of the logs:
```
Starting pihole_cloudflared_1 ... done
Recreating pihole_pihole_1    ... done
Attaching to pihole_cloudflared_1, pihole_pihole_1
cloudflared_1  | 2021-04-15T14:20:06Z INF Adding DNS upstream url=https://1.1.1.1/dns-query
pihole_1       | [s6-init] making user provided files available at /var/run/s6/etc...exited 0.
cloudflared_1  | 2021-04-15T14:20:06Z INF Adding DNS upstream url=https://1.0.0.1/dns-query
cloudflared_1  | 2021-04-15T14:20:06Z INF Starting metrics server on 127.0.0.1:8080/metrics
cloudflared_1  | 2021-04-15T14:20:06Z INF Starting DNS over HTTPS proxy server address=dns://0.0.0.0:5054
pihole_1       | [s6-init] ensuring user provided files have correct perms...exited 0.
pihole_1       | [fix-attrs.d] applying ownership & permissions fixes...
pihole_1       | [fix-attrs.d] 01-resolver-resolv: applying... 
pihole_1       | [fix-attrs.d] 01-resolver-resolv: exited 0.
pihole_1       | [fix-attrs.d] done.
pihole_1       | [cont-init.d] executing container initialization scripts...
pihole_1       | [cont-init.d] 20-start.sh: executing... 
pihole_1       |  ::: Starting docker specific checks & setup for docker pihole/pihole
pihole_1       | Assigning random password: ZiA4pRwf
pihole_1       | 
pihole_1       |   [i] Installing configs from /etc/.pihole...
pihole_1       |   [i] Existing dnsmasq.conf found... it is not a Pi-hole file, leaving alone!
  [✓] Copying 01-pihole.conf to /etc/dnsmasq.d/01-pihole.conf
pihole_1       | Setting DNS servers based on PIHOLE_DNS_ variable
pihole_1       | ::: Pre existing WEBPASSWORD found
pihole_1       | DNSMasq binding to default interface: eth0
pihole_1       | Added ENV to php:
pihole_1       | 			"PHP_ERROR_LOG" => "/var/log/lighttpd/error.log",
pihole_1       | 			"ServerIP" => "192.168.25.233",
pihole_1       | 			"VIRTUAL_HOST" => "192.168.25.233",
pihole_1       | Using IPv4 and IPv6
pihole_1       | ::: Preexisting ad list /etc/pihole/adlists.list detected ((exiting setup_blocklists early))
pihole_1       | https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts
pihole_1       | ::: Testing pihole-FTL DNS: FTL started!
pihole_1       | ::: Testing lighttpd config: Syntax OK
pihole_1       | ::: All config checks passed, cleared for startup ...
pihole_1       | ::: Enabling Query Logging
pihole_1       |   [i] Enabling logging...
  [✓] Logging has been enabled!
pihole_1       |  ::: Docker start setup complete
pihole_1       |   [i] Neutrino emissions detected...
  [✓] Pulling blocklist source list into range
pihole_1       | 
  [✓] Preparing new gravity database
pihole_1       |   [i] Using libz compression
pihole_1       | 
pihole_1       |   [i] Target: https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts
  [✓] Status: Retrieval successful
pihole_1       |   [i] Analyzed 76471 domains
pihole_1       |   [i] List stayed unchanged
pihole_1       | 
  [✓] Storing downloaded domains in new gravity database
  [✓] Building tree
  [✓] Swapping databases
pihole_1       |   [i] Number of gravity domains: 76471 (76471 unique domains)
pihole_1       |   [i] Number of exact blacklisted domains: 0
pihole_1       |   [i] Number of regex blacklist filters: 0
pihole_1       |   [i] Number of exact whitelisted domains: 0
pihole_1       |   [i] Number of regex whitelist filters: 0
  [✓] Cleaning up stray matter
pihole_1       | 
pihole_1       |   [✗] DNS service is NOT listening
pihole_1       | kill: usage: kill [-s sigspec | -n signum | -sigspec] pid | jobspec ... or kill -l [sigspec]
pihole_1       |   Pi-hole version is v5.3.1 (Latest: v5.3.1)
pihole_1       |   AdminLTE version is v5.5 (Latest: v5.5)
pihole_1       |   FTL version is v5.8 (Latest: v5.8)
pihole_1       | [cont-init.d] 20-start.sh: exited 0.
pihole_1       | [cont-init.d] done.
pihole_1       | [services.d] starting services
pihole_1       | Starting pihole-FTL (no-daemon) as root
pihole_1       | Starting crond
pihole_1       | Starting lighttpd
pihole_1       | [services.d] done.
```

While the containers are running, here's a dig from my laptop:
```
kkurz:~ dig +short @192.168.25.233 -p 53 pi-hole.net 
;; connection timed out; no servers could be reached
kkurz:~ dig +short @192.168.25.233 -p 5054 pi-hole.net
192.124.249.118
```

So the upstream I'm providing (local cloudflared DoH container) is responding as expected, but nothing from the PiHole running on :53.

**Comment by uselpa** [comment]
FWIW, I had this issue, and now applied most recommendations from this thread, specifically
- removed DNS1, DNS2 and added PIHOLE_DNS_
- added a backup DNS server for the container (in addition to 127.0.0.1)

and it now appears to work with this compose file

```
version: "3"

services:
  pihole:
    container_name: pihole
    image: pihole/pihole:latest
    ports:
      - "53:53/tcp"
      - "53:53/udp"
#     - "67:67/udp" # DHCP
      - "80:80/tcp"
    environment:
      TZ: 'Europe/Luxembourg'
      PIHOLE_DNS_: 192.168.178.1
    volumes:
      - './v-pihole:/etc/pihole/'
      - './v-dnsmasq.d:/etc/dnsmasq.d/'
    dns:
      - 127.0.0.1
      - 192.168.178.1
    restart: unless-stopped
```

**Comment by flipswitchingmonkey** [comment]
This version works for me now as well:
```
version: "3"

services:
  pihole:
    container_name: pihole
    image: pihole/pihole:latest
    network_mode: host
    environment:
      TZ: 'Europe/Berlin'
      WEBPASSWORD: 'Gyh3Kxyx8ZwQecKH'
      ServerIP: 192.168.178.200
      PROXY_LOCATION: pihole
      VIRTUAL_HOST: pihole.fabrikat.dev
      VIRTUAL_PORT: 80
      WEB_PORT: 6080
    dns:
      - 127.0.0.1
      - 1.1.1.1
    volumes:
      - './etc-pihole/:/etc/pihole/'
      - './etc-dnsmasq.d/:/etc/dnsmasq.d/'
    cap_add:
      - NET_ADMIN
    restart: unless-stopped
```

The only difference was adding the `dns:` section

**Comment by predmijat** [comment]
5.8 works for me too after adding `dns` section to `docker-compose.yml`:

```
dns:
    - 127.0.0.1
    - 1.1.1.1
```

I've read somewhere that first dns entry should be 127.0.0.1 - is that correct?
Also, I'm not sure I quite get the difference between the `dns` section and `PIHOLE_DNS_` in `environment`

**Comment by PromoFaux** [comment]
The environment variable is what Pi-hole FTL uses. So when it receives a query from one of your clients, and it is not on a blocklist, the query will be forwarded to the address(es) stated in `PIHOLE_DNS_`

The `dns` section is to tell the container (or even, the OS within) what DNS server to use to resolve it's own queries

**Comment by PromoFaux** [comment]
So one thing that is for certain here (in my environment)... I can reproduce `dns` section causing issues all the way back to the `:v4.4` container. As soon as I remove it my compose file, the container runs normally. At least on docker on windows. 

I also see exactly the same on my Synology NAS. Once I remove the `dns` section, the container starts normally

I'm still not sure that there is anything actually wrong with the `v5.8` image, per se.

**Comment by squirtbrnr** [comment]
I can confirm that removing the `--dns=127.0.0.1` part from my docker run command allows the container to start and just produces the "DNS service is NOT listening" message in the logs.  I also tried adding `--dns=1.1.1.1` and that also works.  I have not tried leaving `--dns=127.0.0.1` but adding `-e SKIPGRAVITYONBOOT=1` yet as was mentioned earlier.

As I understand from above, the "NOT listening" message is cosmetic and does not actually affect functionality of PiHole currently, aside from the initial startup of a fresh container and the status is "Unknown" and I just need to enable PiHole once.

**Comment by predmijat** [comment]
@squirtbrnr It probably does listen, it's just a race condition. Check the details here: https://discourse.pi-hole.net/t/pi-hole-ftl-v5-8-web-v5-5-and-core-v5-3-released/46237/19

If you do `docker exec -it $your_container_name pihole status` it should say that it's listening.

**Comment by kylekurz** [comment]
@predmijat while that request does return success, I don't get a response from PiHole via a dig request while in that state. So something else is still weird. I do not want to add a `--dns=<some external IP>` as a "fix", I want the internal container to use PiHole as DNS and my DoH upstream to be queried when PiHole doesn't have an answer. As stated in my previous comment, I can get a response from the upstream, so it should not be necessary for me to provide another DNS flag for the OS.

**Comment by kylekurz** [comment]
A little more info here, I removed cloudflare from my docker-compose and am just relying on pihole. Still had issues, so I went all the way down to running the container directly without compose. That worked, without a persistence mount, so I blew away the local data and ran just pihole in compose. I still get the error in startup logs, but I *do* get responses to dig now. Now to try to get cloudflared back.

**Comment by thomasmerz** [comment]
@PromoFaux asked:
> All, just a hunch.. what is the output of `docker --version` on your affected machines?

`Docker version 19.03.15, build 99e3ed89195c` on OpenSUSE 15.2
`Docker version 19.03.8, build afacb8b7f0` on Ubuntu 20.04.2 LTS
`Docker version 19.03.6, build 369ce74a3c` on Ubuntu 18.04.5 LTS

**Comment by humsg** [comment]
> The only way I was able to reproduce any issue was running a single localhost DNS server in the below style, which I think a group of people has an issue with an update or two ago also.
> 
> ```
>     environment:
>       PH_VERBOSE: 1
>     dns:
>       - 127.0.0.1
> 
> ... equals ...
> 
> pihole         | + gravity.sh
> pihole         |   [✗] DNS service is not running
> Looping until failure
> ```
> 
> I think solo localhost dns only works if you have the environment variable `SKIPGRAVITYONBOOT: 1` .
> 
> So that^ or setting a secondary DNS should work for most.

that is true, I think v5.7 can be started without forcing to update the gravity, if I added a second public dns other than my own(which I dont have right now), my dns requests can be exposed when restarting pihole, it is pointless, it's against the reason why I set up my own dns server.

**Comment by Jiheffe** [comment]
> Same issues on x86. Had to revert back to v5.7.

Maybe noob question... How do you revert back? I tried to `pull pihole/pihole:5.7` but that doesn't work. How do you find a list of valid tags por pihole?

**Comment by mdzw** [comment]
I believe the correct way to solve this issue is to change when "Gravity on Boot" is run.

The s6 init system has [different stages](https://github.com/just-containers/s6-overlay#init-stages). Currently, "gravity on boot" is [run](https://github.com/pi-hole/docker-pi-hole/blob/master/s6/debian-root/etc/cont-init.d/20-start.sh#L18) during Stage 2.ii: **cont-init.d**. One instance of pihole-FTL is started during cont-init, but it is only there to check the validity of the config files; it exits soon after starting. The final "service mode" instance of pihole-FTL is not started until **Stage 2.iii**, when the supervisor starts doing work.

If gravity.sh is counting on FTL to do its DNS lookups, then we should not run gravity until the supervised instance of FTL is running. We can accomplish this by **moving "Gravity on Boot" to a `@reboot` line in /etc/cron.d/gravity-on-boot**, and making that file's existence dependent on the value of SKIPGRAVITYONBOOT.

This will work because cron isn't started until we've reached Stage 2.iii.

**Comment by PromoFaux** [comment]
All - I've just merged @mdzw 's propsal into `dev`, and the images have now been built and pushed to docker hub.

Please try running the `:dev` tag to see if the issue is resolved. Report back either way, please!

**Comment by kylekurz** [comment]
@PromoFaux I can confirm I'm no longer seeing the FTL startup errors on :dev. I am still having issues with my Cloudflared setup, but I think that's my own thing, as using Cloudflare directly as upstream (1.1.1.1), I have no problems. Thanks for digging into this @mdzw!

**Comment by GnaphronG** [comment]
I updated with the latest `dev` image from the docker hub, it's working properly.

**Comment by thomasmerz** [comment]
@PromoFaux Same here: positive feedback for `:dev` - Pihole is up and running properly 👍🏻

**Comment by Jiheffe** [comment]
> Please try running the `:dev` tag to see if the issue is resolved. Report back either way, please!
Can confirm `:dev` is running properly 👍

**Comment by stheine** [comment]
confirm. showing `Pi-hole v5.3.1 Web Interface v5.5 FTL v5.8`, with not more updates available.

**Comment by trentonfuller** [comment]
I am also no longer experiencing issues with 5.8 after pulling the `:dev` image from Docker Hub.

**Comment by PromoFaux** [comment]
Great, thanks everyone for the feedback, and thanks @mdzw  for digging into the code for a fix! Will try and get a re-tag of 5.8 out later...

**Comment by PromoFaux** [comment]
v5.8 re-tagged. Have at it!

**Comment by pingadilim** [comment]
Previously on v5.7, I got error when starting PiHole that DNS not resolving, but it could be fixed by adding new dns for pihole service (127.0.0.1 and 1.1.1.1 or 8.8.8.8). In v5.8 that doesn't work.
So, for v5.8 I've been testing the configuration on my docker-compose.yml and it seems the DNS and PIHOLE_DNS_ environment is the culprit that are causing the error and the container always in state of restarting and the console could not be accessed. I will attach the Error Logs and docker-compose.yml that are working for me.

### Additional Info:
- OS: Ubuntu 20.10
- Kernel: 5.8.0-1020-raspi
- Docker version 19.03.13, build 4484c46
- docker-compose version 1.25.0, build unknown
- PiHole Image: latest (v5.8) downloaded on April 19, 2021 at 07.00 UTC

### Not Working docker-compose.yml:
This is the previous docker-compose.yml file that are not working for me.
```
version: "3.6"
services:
  cloudflared:
    container_name: cloudflared
    image: crazymax/cloudflared:latest
    ports:
      - "5053:5053/udp"
      - "49312:49312/tcp"
    environment:
      - TZ=${TZ}
      - "TUNNEL_DNS_UPSTREAM=https://1.1.1.1/dns-query,https://1.0.0.1/dns-query"
    restart: unless-stopped
    networks:
      pihole-net:
        ipv4_address: 10.0.0.

  test-pihole:
    container_name: test-pihole
    image: pihole/pihole:latest
    restart: unless-stopped
    ports:
      - "80:80/tcp"
      - "53:53/tcp"
      - "53:53/udp"
      - "443:443/tcp"
    cap_add:
      - NET_ADMIN
    environment:
#      - DNS1='10.0.0.2#5053'
#      - DNS2=''
      - PIHOLE_DNS_='10.0.0.2#5053;1.1.1.1'
      - IPv6=false
      - TZ=${TZ}
      - DNSMASQ_LISTENING=all
      - WEBPASSWORD=${PIHOLE_PASSWORD}
    volumes:
      - "${USERDIR}/docker/test-pihole/config:/etc/pihole"
      - "${USERDIR}/docker/test-pihole/dnsmasq.d:/etc/dnsmasq.d"
    networks:
      pihole-net:
        ipv4_address: 10.0.0.100
    dns:
      - 127.0.0.1
      - 1.1.1.1

networks:
  pihole-net:
    driver: bridge
    ipam:
      config:
        - subnet: 10.0.0.0/24
```

### Error Logs:
This is the error logs based on above docker-compose-yml
```
test-pihole    | 2021-04-19T07:25:39.014439321Z [s6-init] making user provided files available at /var/run/s6/etc...exited 0.
test-pihole    | 2021-04-19T07:25:39.498152747Z [s6-init] ensuring user provided files have correct perms...exited 0.
test-pihole    | 2021-04-19T07:25:39.509888510Z [fix-attrs.d] applying ownership & permissions fixes...
test-pihole    | 2021-04-19T07:25:39.522816348Z [fix-attrs.d] 01-resolver-resolv: applying...
test-pihole    | 2021-04-19T07:25:39.588895316Z [fix-attrs.d] 01-resolver-resolv: exited 0.
test-pihole    | 2021-04-19T07:25:39.593115794Z [fix-attrs.d] done.
test-pihole    | 2021-04-19T07:25:39.600875501Z [cont-init.d] executing container initialization scripts...
test-pihole    | 2021-04-19T07:25:39.613738155Z [cont-init.d] 20-start.sh: executing...
test-pihole    | 2021-04-19T07:25:39.763410754Z  ::: Starting docker specific checks & setup for docker pihole/pihole
test-pihole    | 2021-04-19T07:25:41.949932301Z
test-pihole    | 2021-04-19T07:25:41.950022578Z   [i] Installing configs from /etc/.pihole...
test-pihole    | 2021-04-19T07:25:41.960349378Z   [i] Existing dnsmasq.conf found... it is not a Pi-hole file, leaving alone!
test-pihole    | 2021-04-19T07:25:41.967546389Z   [i] Copying 01-pihole.conf to
  [✓] Copying 01-pihole.conf to /etc/dnsmasq.d/01-pihole.conf
test-pihole    | 2021-04-19T07:25:43.181992853Z Setting DNS servers based on PIHOLE_DNS_ variable
test-pihole    | 2021-04-19T07:25:43.183590664Z Invalid IP detected in PIHOLE_DNS_: '10.0.0.2#5053
test-pihole    | 2021-04-19T07:25:43.185102661Z Invalid IP detected in PIHOLE_DNS_: 1.1.1.1'
test-pihole    | 2021-04-19T07:25:43.185208308Z No Valid IPs dectected in PIHOLE_DNS_. Aborting
test-pihole    | 2021-04-19T07:25:43.189622024Z [cont-init.d] 20-start.sh: exited 1.
test-pihole    | 2021-04-19T07:25:43.251534865Z [cont-finish.d] executing container finish scripts...
test-pihole    | 2021-04-19T07:25:43.259954618Z [cont-finish.d] done.
test-pihole    | 2021-04-19T07:25:43.264963474Z [s6-finish] waiting for services.
test-pihole    | 2021-04-19T07:25:43.497254644Z [s6-finish] sending all processes the TERM signal.
test-pihole    | 2021-04-19T07:25:46.519934499Z [s6-finish] sending all processes the KILL signal and exiting.
```

From error logs I found out that the DNS IP being set is what causing the container not being started and s6 sending kill signal so the container will be restarted again. With that I try to comment out all the DNS environment from docker-compose.yml and it works!

### Working docker-compose.yml:
```
version: "3.6"
services:
  cloudflared:
    container_name: cloudflared
    image: crazymax/cloudflared:latest
    ports:
      - "5053:5053/udp"
      - "49312:49312/tcp"
    environment:
      - TZ=${TZ}
      - "TUNNEL_DNS_UPSTREAM=https://1.1.1.1/dns-query,https://1.0.0.1/dns-query"
    restart: unless-stopped
    networks:
      pihole-net:
        ipv4_address: 10.0.0.

  test-pihole:
    container_name: test-pihole
    image: pihole/pihole:latest
    restart: unless-stopped
    ports:
      - "80:80/tcp"
      - "53:53/tcp"
      - "53:53/udp"
      - "443:443/tcp"
    cap_add:
      - NET_ADMIN
    environment:
#      - DNS1='10.0.0.2#5053'
#      - DNS2=''
#      - PIHOLE_DNS_='10.0.0.2#5053;1.1.1.1'
      - IPv6=false
      - TZ=${TZ}
      - DNSMASQ_LISTENING=all
      - WEBPASSWORD=${PIHOLE_PASSWORD}
    volumes:
      - "${USERDIR}/docker/test-pihole/config:/etc/pihole"
      - "${USERDIR}/docker/test-pihole/dnsmasq.d:/etc/dnsmasq.d"
    networks:
      pihole-net:
        ipv4_address: 10.0.0.100
    dns:
#      - 127.0.0.1
      - 1.1.1.1

networks:
  pihole-net:
    driver: bridge
    ipam:
      config:
        - subnet: 10.0.0.0/24
```

Please pay attention on environment data that I commented out all DNS (DNS1 and DNS2 being deprecated and being replace by PIHOLE_DNS_ in the latest PiHole) because what causing error on my system is the DNS' environments. So after the container started, please make sure that the custom DNS inside Settings > DNS already being set.

Hope this post could assist anyone with the same problem and will help development team to rechecked what causing the problems.

**Comment by PromoFaux** [comment]
@pingadilim - I've added some notes to the readme for v5.8:

https://github.com/pi-hole/docker-pi-hole#v58

TL;DR, change the way you're passing the environment variables:

`- PIHOLE_DNS_='10.0.0.2#5053;1.1.1.1'` is invalid, whereas `- PIHOLE_DNS_=10.0.0.2#5053;1.1.1.1` would be valid, for example.

**Comment by pingadilim** [comment]
> @pingadilim - I've added some notes to the readme for v5.8:
> 
> https://github.com/pi-hole/docker-pi-hole#v58
> 
> TL;DR, change the way you're passing the environment variables:
> 
> `- PIHOLE_DNS_='10.0.0.2#5053;1.1.1.1'` is invalid, whereas `- PIHOLE_DNS_=10.0.0.2#5053;1.1.1.1` would be valid, for example.

Noted @PromoFaux Thanks for the additional notes!
